ibis/sql/postgres/tests/test_udf.py

```diff
@@ -0,0 +1,173 @@
+"""Test support for already-defined UDFs in Postgres"""
+
+import random
+import string
+
+import pytest
+
+import pandas as pd
+import ibis.expr.datatypes
+import ibis.sql.postgres.udf.api
+from ibis.sql.postgres.client import PostgreSQLClient
+
+
+datatypes = ibis.expr.datatypes
+
+# mark test module as postgresql (for ability to easily exclude,
+# e.g. in conda build tests)
+# (Temporarily adding `postgis` marker so Azure Windows pipeline will exclude
+#     pl/python tests.
+#     TODO: update Windows pipeline to exclude postgres_extensions
+#     TODO: remove postgis marker below once Windows pipeline updated
+pytestmark = [
+    pytest.mark.postgresql,
+    pytest.mark.postgis,
+    pytest.mark.postgres_extensions,
+]
+
+# Database setup (tables and UDFs)
+
+
+test_schema = None
+
+
+def gen_schema_name(basename='test'):
+    """Generate a random alpha string starting with 'test_' to be used as a
+    test schema name"""
+    schema_name = '{}_{}'.format(
+        basename,
+        ''.join([random.choice(string.ascii_lowercase) for i in range(6)])
+    )
+    return schema_name
+
+
+def create_test_schema(connection: PostgreSQLClient, max_tries=20):
+    """Create a random test schema, necessary to allow for parallel pytest
+    testing"""
+    schema_exists = True
+    tries = 0
+    while schema_exists and tries < max_tries:
+        new_name = gen_schema_name()
+        sql_check_schema = """SELECT count(schema_name) as n_recs
+        FROM information_schema.schemata
+        WHERE lower(schema_name) = lower('{}');""".format(new_name)
+        df_result = pd.read_sql(sql_check_schema, connection.con)
+        schema_exists = df_result['n_recs'].iloc[0] > 0
+        tries += 1
+    assert tries > 0
+    global test_schema
+    test_schema = new_name
+    connection.con.execute("CREATE SCHEMA {};".format(new_name))
+
+
+table_name = 'udf_test_users'
+
+sql_table_setup = """DROP TABLE IF EXISTS {schema}.{table_name};
+CREATE TABLE {schema}.{table_name} (
+    user_id integer,
+    user_name varchar,
+    name_length integer
+);
+INSERT INTO {schema}.{table_name} VALUES
+(1, 'Raj', 3),
+(2, 'Judy', 4),
+(3, 'Jonathan', 8)
+;
+"""
+
+sql_define_py_udf = """CREATE OR REPLACE FUNCTION {schema}.pylen(x varchar)
+RETURNS integer
+LANGUAGE plpythonu
+AS
+$$
+return len(x)
+$$;"""
+
+sql_define_udf = """CREATE OR REPLACE FUNCTION {schema}.custom_len(x varchar)
```

Comment:
Make all of these module globals into `pytest.fixture`s, it's easier to see what's being used and where it's being used.